server/src/main/java/org/apache/druid/curator/CuratorModule.java

```diff
@@ -97,6 +99,21 @@ public CuratorFramework makeCurator(ZkEnablementConfig zkEnablementConfig, Curat
 
     framework.getUnhandledErrorListenable().addListener((message, e) -> {
       log.error(e, "Unhandled error in Curator, stopping server.");
+      final Thread halter = new Thread(
+          () -> {
+            try {
+              Threads.sleepFor(10, TimeUnit.SECONDS);
```

Comment:
At first I thought 10s seemed short. But now I'm not so sure. I suppose it's fine. We're trying to shut down the server quickly; the only reason we're waiting any time at all is to give shutdown hooks a chance to run, clean up temporary files, close upstream connections cleanly, finish currently-executing queries, etc.

Due to the "finish currently-executing queries" thing, maybe we should go with 30s?

Reply:
Generally speaking, this failure mode should only happen at startup.  I suppose it's possible for unhandled background exceptions to make it out later on in the lifetime of the process, but I'm only familiar with these happening at startup and indicating problems with ZK connectivity/configurations.  As such, I don't think that "finish currently executing queries" is really a thing.  That said, if you feel strongly about it, 30s is better than nothing existing so I'll do it to get it merged :)